fix(Core/Trainer): Fixes display of ranks in Trainer (Horn of Winter and similar)

EricksOliveira opened the pull request:
Co-authored-by: Shauren <shauren.trinity@gmail.com>

https://github.com/TrinityCore/TrinityCore/commit/baebb2d602ca2f10159ea0039ec3a6d03155c3bd

Implements proper sending of the SMSG_SEND_UNLEARN_SPELLS packet to correct the display of spell ranks on NPC Trainers.

Before this change, when training the second rank of Horn of Winter, the first rank remained listed in the Trainer until the window was closed and reopened, causing visual inconsistency.

The changes introduce:

Player::SendUnlearnSpells to send to the client spells that should no longer be displayed.

Player::IsUnlearnNeededForSpell to check if a given spell is part of a rank chain and needs to be handled in the packet.

This means the Trainer now correctly hides previous ranks after learning a new one, ensuring expected behavior for chained spells.

<!-- First of all, THANK YOU for your contribution. -->

## Changes Proposed:
<!-- If your pull request promotes complex changes that require a detailed explanation, please describe them in detail specifying what your solution is and what is it meant to address. -->
This PR proposes changes to:
-  [X] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:
<!-- If your fix has a relating issue, link it below -->
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/4590

## SOURCE:
<!-- If you can, include a source that can strengthen your claim -->
The changes have been validated through:
- [X] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [X] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:
<!-- Does it build without errors? Did you test in-game? What did you test? On which OS did you test? Describe any other tests performed -->
This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.


## How to Test the Changes:
<!-- Describe in a detailed step-by-step order how to test the changes -->

- [X] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:
<!-- Is there anything else left to do after this PR? -->

- [ ]
- [ ]

<!-- If you intend to contribute repeatedly to our project, it is a good idea to join our discord channel. We set ranks for our contributors and give them access to special resources or knowledge: https://discord.com/invite/GyFvXpk7)
     Do not remove the instructions below about testing, they will help users to test your PR -->
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

EricksOliveira committed:
fix(Core/Player): Fixes visual bug in Death Knight trainer and other classes for spells with multiple ranks

The fix was implemented generically in the Player::GetTrainerSpellState() function, ensuring that, for any spell that is part of a rank chain, the trainer correctly recognizes whether the player already knows any higher rank and marks the spell as "already learned" (TRAINER_SPELL_GRAY).

This improvement prevents the visual bug not only for Death Knights, but for any class with spells at multiple ranks, aligning the behavior with what's expected in the official client.

Rorschach91 commented:
> This seems to work. <img alt="work" width="619" height="782" src="https://private-user-images.githubusercontent.com/108557877/477229221-19d7eedc-81f4-4fbb-bfa0-07b530596204.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUwMzIyNjUsIm5iZiI6MTc1NTAzMTk2NSwicGF0aCI6Ii8xMDg1NTc4NzcvNDc3MjI5MjIxLTE5ZDdlZWRjLTgxZjQtNGZiYi1iZmEwLTA3YjUzMDU5NjIwNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxMlQyMDUyNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZDhlMjk4YTlhN2Y0YWE3MzdjYmEyOGFjMjFjMmU3ZmRhYzhhZGZjZDRkNzQ0YTM5Nzg2NWQyNmNkMzFmODVkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3P6L3tNvFXPQswm2aiLAg6eeq8N2QDIwmR-pLJsk5KQ">

Edit: it seems that this solution causes more problems.

EricksOliveira committed:
Possible fix

EricksOliveira committed:
Merge branch 'master' into patch-9

EricksOliveira committed:
fix

EricksOliveira committed:
Fix

EricksOliveira committed:
Fix comment

EricksOliveira commented:
> @EricksOliveira are you able do a quick resume what does this fix and how to test it?

Previously, the Rank 1 Spell Horn of Winte would reappear after training, but the Rank 2 Spell Horn wouldn't appear. This no longer occurs. A requirement was introduced for the Rank 2 Spell Horn to only appear when the Rank 1 Spell Horn is trained.

EricksOliveira committed:
Core/Spells: Fixes display of ranks in Trainer (Horn of Winter and similar)

Implements proper sending of the SMSG_SEND_UNLEARN_SPELLS packet to correct the display of spell ranks on NPC Trainers.

Before this change, when training the second rank of Horn of Winter, the first rank remained listed in the Trainer until the window was closed and reopened, causing visual inconsistency.

The changes introduce:

Player::SendUnlearnSpells to send to the client spells that should no longer be displayed.

Player::IsUnlearnSpellsPacketNeededForSpell to check if a given spell is part of a rank chain and needs to be handled in the packet.

This means the Trainer now correctly hides previous ranks after learning a new one, ensuring expected behavior for chained spells.

EricksOliveira commented:
> Fresh DK, go see trainer, only rank 1 is there, learn it, rank 2 appears and rank 1 is gone, learn rank 2, rank 1 reappears, if you close trainer window and reopen it, rank 1 is now gone. The PR does seem to fix the issue at hand, least at first glance, you can learn both rank now and everything is in order, but its creating a new bug where rank 1 reappears at the end after learning rank 2. So I would say the issue remains. Here's a demo. [Video] https://www.youtube.com/watch?v=hK4h9Jv0c9o

Thanks for the feedback. The fix was only partially implemented, and I was still investigating. But I managed to uncover the real issue. The fix has been submitted.

heyitsbench commented on src/server/game/Handlers/NPCHandler.cpp:
```diff
@@ -264,6 +264,10 @@ void WorldSession::HandleTrainerBuySpellOpcode(WorldPacket& recvData)
     else
         _player->learnSpell(spellId);
 
+    // Send "unlearn" packet only if necessary
```
Comment redundant.

EricksOliveira replied:
Ok

heyitsbench commented on src/server/game/Entities/Player/Player.h:
```diff
@@ -1656,6 +1656,9 @@ class Player : public Unit, public GridObject<Player>
     void AddNewMailDeliverTime(time_t deliver_time);
 
+    void SendUnlearnSpells();
+    static bool IsUnlearnSpellsPacketNeededForSpell(uint32 spellId);
```
```suggestion
    static bool IsUnlearnNeededForSpell(uint32 spellId);
```
Not a fan of the name being as long as it is, feels unnecessary.

heyitsbench commented:
Please attach the commit and co-author where this came from.

EricksOliveira committed:
Fix comment and big name

EricksOliveira committed:
Merge branch 'master' into patch-9

EricksOliveira committed:
Fix: Prevent lower-rank spell override when learning new abilities (Unlearn packet guard)

The solution adds a generic check called Player::IsUnlearnNeededForSpell() that prevents the unnecessary sending of the SMSG_SEND_UNLEARN_SPELLS packet during normal spell learning.

The system now only sends the unlearn packet when truly necessary—for example, in cases of rank changes within the same spell chain.

Key points:

Added generic IsUnlearnNeededForSpell() function.

Removed unconditional sending of SendUnlearnSpells() when learning spells.

Fixed the issue without hardcoding, maintaining compatibility with any spell in the game.

Prevents incorrect client behavior when training NPC trainer skills.

Maintains the normal learning flow and rank progression.

Co-Authored-By: Shauren <shauren.trinity@gmail.com>

EricksOliveira committed:
Fix comment

Co-Authored-By: Shauren <shauren.trinity@gmail.com>

EricksOliveira commented:
New fix. Awaiting testing.

Kitzunu committed:
Merge branch 'master' into patch-9

Kitzunu committed:
Merge branch 'master' into patch-9

EricksOliveira committed:
Merge branch 'master' into patch-9

EricksOliveira committed:
Merge branch 'master' into patch-9

EricksOliveira committed:
unlearn spell packet cleared by empty send on login

The populated SMSG_SEND_UNLEARN_SPELLS packet sent from SendInitialSpells()
was immediately overwritten by an empty SMSG_SEND_UNLEARN_SPELLS packet
still being sent from SendInitialPacketsBeforeAddToMap() right after,
causing the unlearn-eligible spells to never actually persist on login.

- Move the unlearn packet send out of SendInitialSpells() and call
  Player::SendUnlearnSpells() from SendInitialPacketsBeforeAddToMap()
  instead, replacing the leftover empty packet.
- Add a SkillLineAbilityEntry::SupercededBySpell check in both
  SendUnlearnSpells() and IsUnlearnNeededForSpell(), since the client
  already hides ranks that have a superseding rank on its own. Without
  this check, superseded ranks were sent redundantly, causing the
  "shows briefly and disappears" flicker.

Ported from TrinityCore commit baebb2d.

Co-authored-by: Shauren <shauren.trinity@gmail.com>

sogladev reviewed (approved):
✅ I wasn't able to reproduce the issue anymore by doing https://github.com/azerothcore/azerothcore-wotlk/pull/22678#issuecomment-3338708807c
